bazel, ci: use test2json format for nightly logging

rail opened the pull request:
Fixes #71930

Previously, bazel based nightlies didn't report errors as github issues.

This PR uses a custom feature of `rules_go` to save the log output in
`test2json` format. A new shell function is added to post-process the
generated log and ports the `run_json_test` features to submit failures
to GitHub.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/72611)
<!-- Reviewable:end -->

rickystewart commented on build/patches/com_github_bazelbuild_rules_go.patch:
```diff
@@ -0,0 +1,25 @@
+ 			return fmt.Errorf("error while generating testreport: %s", werr)
+ 		}
+ 	}
++	if out, ok := os.LookupEnv("GO_TEST_JSON_OUTPUT_FILE"); ok {
```
We already have a fork of `rules_go` with some patches applied, can you just push this patch to our repo instead? We update `rules_go` not infrequently, and keeping this patch file up-to-date is going to be way more complicated than using Git.

1. Check out `release-0.26` of `cockroachdb/rules_go` (should be commit `4a7bcc9b9051eb6a12bcb03a9bf38138c4bbc2ea`)
2. Push this commit to `release-0.26`
3. Download the new tarball: `github.com/cockroachdb/rules_go/tarball/$NEW_COMMIT_ID`
4. Mirror the file to `https://storage.googleapis.com/public-bazel-artifacts/bazel/` and update `WORKSPACE` accordingly. There shouldn't be a need to apply the patch to `DEPS.bzl` so you can leave it untouched.

rail replied:
Good point. Probably I can just use my own fork for faster iterations instead of using this method.

rail committed:
bazel, ci: use test2json format for nightly logging

Fixes #71930

Previously, bazel based nightlies didn't report errors as github issues.

This PR uses a custom feature of `rules_go` to save the log output in
`test2json` format. A new shell function is added to post-process the
generated log and ports the `run_json_test` features to submit failures
to GitHub.

Release note: None

rail commented:
I think this is ready to be reviewed.

I pushed [this rules_go change](https://github.com/cockroachdb/rules_go/commit/1ce9e349278b6e9ea1d52e0f73446c70ad440cbd) to the [release-0.29](https://github.com/cockroachdb/rules_go/commits/release-0.29) branch, so we can save logs in `test2json` format.

To be tested (in production): `github-post` requires `GITHUB_TOKEN` in order to work, I'll monitor the builder.

This is how it behaves when there is a test failure: [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_ScratchProjectPutTcExperimentsInHere_RailSqlLogicTestHighVModuleNightlyBazel/3798226?buildTab=artifacts#%2Ffull_output.tgz)
<img width="392" alt="image" src="https://user-images.githubusercontent.com/42865/143471302-30361837-634a-4df8-b692-25a4863b2544.png">

And there is also a green test: [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_ScratchProjectPutTcExperimentsInHere_RailSqlLogicTestHighVModuleNightlyBazel/3798756?buildTab=artifacts)

rail commented:
bors r=rickystewart
